File: 48.Rotate-Image.java
// https://leetcode.com/problems/rotate-image/
//
// algorithms
// Medium (48.67%)
// Total Accepted: 254,139
// Total Submissions: 522,212
// beats 100.0% of java submissions

public class Solution {
    public void rotate(int[][] matrix) {
        int len = matrix.length;
        if(len <= 1) return ;
        int temp [] = new int [2 * len];
        for(int i = 0; i < len / 2; i++) {
            temp[1] = matrix[i][i];
            matrix[i][i] = matrix[len - 1 - i][i];
            temp[2] = matrix[i][len - 1 - i];
            matrix[i][len - 1 - i] = temp[1];
            temp[1] = matrix[len - 1 - i][len - 1 - i];
            matrix[len - 1 - i][len - 1 - i] = temp[2];
            matrix[len - 1 - i][i] = temp[1];
        }
        for(int i = 0; i < len / 2; i++) {
            int num1 = 0;
            int num2 = len;
            for(int j = i + 1; j <= len - 1 - i - 1; j++) {
                temp[num1] = matrix[i][j];
                matrix[i][j] = matrix[len - 1 - j][i];
                num1 += 1;
                temp[num2] = matrix[len - 1 - i][j];
                matrix[len - 1 - i][j] = matrix[len - 1 - j][len - 1 - i];
                num2 += 1;
            }
            int n1 = 0;
            int n2 = len;
            for(int j = i + 1; j <= len - 1 - i - 1; j++) {
                matrix[j][len - 1 - i] = temp[n1];
                n1 += 1;
                matrix[j][i] = temp[n2];
                n2 += 1;
            }
        }
        return ;
    }
}

File: 48.Rotate-Image.py
# https://leetcode.com/problems/rotate-image/
#
# algorithms
# Medium (50.0%)
# Total Accepted:    277,919
# Total Submissions: 555,398


class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        
        for i in xrange((n + 1) / 2):
            l, r = i, n - 1 - i
            t, b = l, r
            if l == r:
                break
            
            tmp = matrix[t][r]
            matrix[t][r] = matrix[t][l]
            tmp1 = matrix[b][r]
            matrix[b][r] = tmp
            matrix[t][l] = matrix[b][l]
            matrix[b][l] = tmp1
            
            tmp = []
            for i in xrange(t + 1, b):
                tmp.append(matrix[i][r])
                matrix[i][r] = matrix[t][i]
            
            tmp1 = []
            for i in xrange(r - 1, l, -1):
                tmp1.append(matrix[b][i])
                matrix[b][i] = tmp[r - 1 - i]
            
            for i in xrange(l + 1, r):
                matrix[t][i] = matrix[b - (i - l)][l]
            
            for i in xrange(b - 1, t, -1):
                matrix[i][l] = tmp1[b - 1 - i]
        
        return
            